File: EE 5561 Notes/Files/Homeworks/Homework 3/Assignment3Files/old/cg_solve.py
import numpy as np

def cg_solve(b, A, no_iter=3):
    
# ----cg_solve------------------------------------------
#     x_recon = A^-1.b
#     no_iter: numver of iterations
#     x_recon = cg_solve(b,A,no_iter)

    x0 = np.zeros_like(b)
    x = np.copy(x0)
    r = b - A(x)
    p = np.copy(r)
    rsold = np.sum(r * np.conj(r))
    
    for ind in range(no_iter):
        Ap = A(p)
        alpha = rsold/np.sum(p * Ap)
        x = x + alpha * p
        r = r - alpha * Ap
        rsnew = np.sum(r * np.conj(r))
        p = r + rsnew / rsold * p
        rsold = np.copy(rsnew)
    return x

File: EE 5561 Notes/Files/Homeworks/Homework 3/Assignment3Files/cg_solve.m
function x = cg_solve(b, A, no_iter)

if nargin < 3
    no_iter = 3;
end

x0 = zeros(size(b), 'single');
x = x0;
r=b-A(x);
p=r;
rsold=r(:)'*r(:);

for ind = 1:no_iter
    Ap = A(p);
    alpha = rsold/(p(:)'*Ap(:));
    x = x + alpha*p;
    r = r - alpha*Ap;
    rsnew = r(:)'*r(:);
    p = r + rsnew/rsold*p;
    rsold = rsnew;
end